File: 1578. Minimum Time to Make Rope Colorful.cpp
// O(n) time and O(1) space complexity
class Solution {
public:
    int minCost(string colors, vector<int>& neededTime) {
        int timeSpent = 0;
        int prevIdx = 0;
        char prevColor = colors[0];
        int sumTimeNeeded = 0;
        int maxTimeNeeded = neededTime[0];

        for (int i = 1; i < colors.size(); i++) {

            if (colors[i] != prevColor) {
                if (i - prevIdx > 1) {
                    timeSpent += sumTimeNeeded;
                };
                prevIdx = i;
                prevColor = colors[i];
                sumTimeNeeded = 0;
                maxTimeNeeded = neededTime[i];

            } else {
                sumTimeNeeded += min(maxTimeNeeded, neededTime[i]);
                maxTimeNeeded = max(maxTimeNeeded, neededTime[i]);
            };
        };

        if (colors.size() - prevIdx > 1) {
            return timeSpent + sumTimeNeeded;
        };
        return timeSpent;
    };
};

File: 1578. Minimum Time to Make Rope Colorful.py
# Another solution
# O(n) time and O(1) space complexity
class Solution:
    def minCost(self, colors: str, neededTime: List[int]) -> int:
        timeSpent = 0
        prevIdx = 0
        prevColor = colors[0]
        sumTimeNeeded, maxTimeNeeded = 0, neededTime[0]

        for i, [color, time] in zip(
            range(1, len(colors)), 
            zip(
                colors[1:],
                neededTime[1:]
                )
            ):

            if color != prevColor:
                if i - prevIdx > 1:
                    timeSpent += sumTimeNeeded
                prevIdx = i
                prevColor = color
                sumTimeNeeded, maxTimeNeeded = 0, time

            else:
                sumTimeNeeded += min(maxTimeNeeded, time)
                maxTimeNeeded = max(maxTimeNeeded, time)

        if len(colors) - prevIdx > 1:
            return timeSpent + sumTimeNeeded
        return timeSpent

# # My first solution:
# # O(n) time and O(1) space complexity
# class Solution:
#     def minCost(self, colors: str, neededTime: List[int]) -> int:
#         timeSpent = 0
#         prev = 0
#         currentColor = colors[0]

#         for i, color in zip(range(1, len(colors)), colors[1:]):
#             if color != currentColor:
#                 if i - prev > 1:
#                     timeSpent += sum(neededTime[prev:i]) - max(neededTime[prev:i])
#                 prev = i
#                 currentColor = color

#         if len(colors) - prev > 1:
#             timeSpent += sum(neededTime[prev:]) - max(neededTime[prev:])

#         return timeSpent
